README.rst

```diff
@@ -348,39 +350,98 @@ A list of sequentially linked nodes. This allows efficient insertion of nodes to
+Finally, SortedMultiDict is similar to SortedDict except that each key
+can be associated with multiple values.  The (key,value) pairs in
+a SortedMultiDict are stored according to the sorted order for keys,
+and (key,value) pairs with the same
+key are stored in order of insertion. 
+
+The containers internally use a 2-3 tree, which is a
 kind of balanced tree and is described in many elementary data
-structure textbook.
+structure textbooks.
 
-This container requires two functions to compare keys: a *less-than* and
+The containers require two functions to compare keys: a *less-than* and
 *equals* function.  With the
 default ordering argument, the comparison
 functions are ``isless(a,b)`` and ``isequal(a,b)`` where ``a`` and ``b``
 are keys.
-User-specified ordering functions are discussed below.
+More details are provided below.
 
 ------------------------------
 Tokens for Sorted Containers
 ------------------------------
 
-The SortedDict type is accompanied by an auxiliary type called the *token*
-and is defined as type ``SDToken``.  A token is an item that stores
-the address of a single data item in the SortedDict and can be
+The sorted container objects are accompanied by auxiliary objects
+called *tokens*
+defined as two-entry tuples and aliased as
+``SDToken``, ``SMDToken``, and ``SetToken`` for SortedDict, SortedMultiDict
+and SortedSet respectively.
+A token is the
+address of a single data item in the container and can be
 dereferenced in time O(1).
-This notion of token is similar to the concept of iterators used
+
+The first entry of a Token tuple is the container as a whole, and the
+second refers to the particular item.  The second part is called a
+*semitoken*.  The types for a semitoken are ``SDSemiToken``, ``SMDSemiToken``,
+and ``SetSemiToken`` for the three types of containers SortedDict, SortedMultiDict
+and SortedSet.  These types are
+all aliases of ``IntSemiToken``.
+
+A restriction for the sorted containers is that
+``IntSemiToken`` or its aliases cannot used as the key-type.  
+This is because
+ambiguity would result between the
+two subscripting calls ``m[k]`` and ``m[st]`` described below.  In the
+rare scenario that a sorted container whose key-type is ``IntSemiToken``
+is required,
+a workaround is to wrap the key inside another immutable
+structure.
+
+In the current
+version of Julia, it is costly to operate on tuples whose entries are not bits-types
+because such tuples are allocated on the heap.
+For example, the first entry of a token is a pointer to a container
+(a non-bits type), so
+a new token is allocated on the heap rather than the
+stack.  In order to avoid performance loss, the package uses tokens
+less frequently than semitokens.  For a function 
+taking a token as an argument like
+``deref`` described below, if it is invoked by explicitly naming
+the token like this::
+
+    t = (m,s)
+    k,v = deref(t)
+
+then there may be a loss of performance
+compared to::
+
+    k,v = deref((m,s))
```

Comment:
While I remember the discussion where you found this to be true, I'm find this description challenging to understand.  In particular, this syntax seems to indicate that I can pass a tuple `(m,s)` to `deref`, yet I don't see that function anywhere--did I miss it?  Can you point to the original discussion where you discovered this?